### PR TITLE
feat: product commission

### PR DESCRIPTION
**Pull request by obi-ewan**
Enable product operators building on the protocol to begin charging commission on settlement.

The new field on `market_position` - `matched_risk_per_product: Vec<ProductMatchedRiskAndRate>`, records the proportions of matched risk each product contributed to a user's overall market position. At settlement, these proportions are used to deduct the relevant fee `rate` from the `market_position` profit.

e.g. 
In the following scenario:
1.  A `market_position` has $100 total profit
2. ProductA charges 1%  commission, and handled 50% of the risk
3. ProductB charges 2%  commission, and handled 50% of the risk

The commission deductions would be:
- ProductA is eligible to take 1% from 50% of the profit - $0.5
- ProductB is eligible to take 2% from 50% of the profit - $1
- The remainder is still paid directly to the user - $98.5

**Potential Breaking Changes**

_**Instructions:**_

Added `commission_payment_queue` account to instructions: `CreateMarket`, `SettleMarketPosition`, `CloseMarket`

_**Accounts**:_

Order (`order_account`)
- Added `product_commission_rate: f64`
- Changed `product: Pubkey` is now product: `Option<Pubkey>`

Market Position (`market_position_account`)
- Added `matched_risk: u64`
- Added `matched_risk_per_product: Vec<ProductMatchedRiskAndRate>`
    - `ProductMatchedRiskAndRate`: `{ product: Pubkey, risk: u64, rate: f64 }`

### CODE REVIEW

File: programs/monaco_protocol/src/state/order_account.rs [hunk 1]
```diff
@@ -28,26 +28,28 @@ pub struct Order {
...
     pub creation_timestamp: i64,
     pub delay_expiration_timestamp: i64,
     // matching data
-    pub stake_unmatched: u64, // stake amount available for matching
-    pub payout: u64,          // amount paid to purchaser during settlement for winning orders
-    pub payer: Pubkey,        // solana account fee payer
+    pub stake_unmatched: u64,         // stake amount available for matching
+    pub payout: u64, // amount paid to purchaser during settlement for winning orders
+    pub payer: Pubkey, // solana account fee payer
+    pub product_commission_rate: f64, // product commission rate at time of order creation
```
Comment (eoin-betdex) [inline programs/monaco_protocol/src/state/order_account.rs:41]: I don't see any client change to add this field to the client types. Is this intentional?

File: programs/monaco_protocol/src/state/order_account.rs [hunk 2]
```diff
@@ -28,26 +28,28 @@ pub struct Order {
     pub market_outcome_index: u16, // market outcome on which order was made
     pub for_outcome: bool, // is order for or against the outcome
     pub order_status: OrderStatus, // status
-    pub product: Pubkey,   // product this order was placed on
+    pub product: Option<Pubkey>, // product this order was placed on
```
Comment (eoin-betdex) [inline programs/monaco_protocol/src/state/order_account.rs:31]: The client type for this field will need to be updated. Currently it is
```
product: PublicKey;
```

File: programs/monaco_protocol/src/state/market_position_account.rs [hunk 3]
```diff
@@ -12,6 +12,20 @@ pub struct MarketPosition {
     pub market_outcome_sums: Vec<i128>,
     pub outcome_max_exposure: Vec<u64>,
     pub payer: Pubkey, // solana account fee payer
+    pub matched_risk: u64,
+    pub matched_risk_per_product: Vec<ProductMatchedRiskAndRate>,
```
Comment (eoin-betdex) [inline programs/monaco_protocol/src/state/market_position_account.rs:16]: Should these be available in the client?

File: npm-admin-client/src/market_helpers.ts [hunk 4]
```diff
@@ -158,6 +158,37 @@ export async function findEscrowPda(
...
+ * @param program {program} anchor program initialized by the consuming client
+ * @param marketPk {PublicKey} publicKey of a market
+ * @returns {FindPdaResponse} PDA of the payment queue
+ *
+ * @example
+ *
+ * const marketPk = new PublicKey('7o1PXyYZtBBDFZf9cEhHopn2C9R4G6GaPwFAxaNWM33D')
+ * const commissionQueuePda = await findEscrowPda(program, marketPK)
+ */
+export async function findMarketCommissionQueuePda(
```
Comment (eoin-betdex) [inline npm-admin-client/src/market_helpers.ts:173]: Nitpick: we use `commission_payments_queue` within the protocol codebase. Should this be a little more like that?

**Review by eoin-betdex — COMMENTED** [review]
Just some surface level review comments.

File: npm-admin-client/src/market_helpers.ts (hunk 4 above)
Comment (obi-ewan) [reply npm-admin-client/src/market_helpers.ts:173]: Was just trying to keep it a bit more concise, but I can update to be more consistent

Comment (eoin-betdex) [reply npm-admin-client/src/market_helpers.ts:173]: It's a nitpick, either way is fine, though usually I'd err on the side of consistency.

**Review by pawel-betdex — DISMISSED** [review]

File: programs/monaco_protocol/src/state/market_position_account.rs (hunk 3 above)
Comment (pawel-betdex) [reply programs/monaco_protocol/src/state/market_position_account.rs:16]: I do not think so. Only modifiable by Monaco itself.

**Review by pawel-betdex — APPROVED** [review]

**Review by eoin-betdex — APPROVED** [review]

**Pull request merged**